File: backend/app/models/raid.py
import uuid
from datetime import date
from decimal import Decimal

from sqlalchemy import ForeignKey, Numeric
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base
from app.models.mixins import TimestampColumns, UUIDPrimaryKey


class RiskLog(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "risk_log"

    risk_code: Mapped[str] = mapped_column(unique=True)
    project_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("projects.id", ondelete="CASCADE"))

    risk_title: Mapped[str]
    risk_description: Mapped[str | None]
    risk_category: Mapped[str | None]
    risk_type: Mapped[str | None]
    identified_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    identified_date: Mapped[date | None]
    risk_owner: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    trigger_event: Mapped[str | None]
    probability: Mapped[str | None]
    impact: Mapped[str | None]
    risk_score: Mapped[int | None]
    severity: Mapped[str | None]
    affected_deliverables: Mapped[str | None]
    affected_milestone: Mapped[str | None]
    response_strategy: Mapped[str | None]
    mitigation_plan: Mapped[str | None]
    contingency_plan: Mapped[str | None]
    residual_risk: Mapped[str | None]
    target_resolution_date: Mapped[date | None]
    current_status: Mapped[str]
    escalation_required: Mapped[bool]
    escalated_to: Mapped[str | None]
    last_review_date: Mapped[date | None]
    next_review_date: Mapped[date | None]
    closure_date: Mapped[date | None]
    remarks: Mapped[str | None]


class IssueLog(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "issue_log"

    issue_code: Mapped[str] = mapped_column(unique=True)
    project_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("projects.id", ondelete="CASCADE"))

    issue_title: Mapped[str]
    issue_description: Mapped[str | None]
    issue_category: Mapped[str | None]
    priority: Mapped[str | None]
    severity: Mapped[str | None]
    raised_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    raised_date: Mapped[date | None]
    assigned_to: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    root_cause: Mapped[str | None]
    business_impact: Mapped[str | None]
    affected_deliverables: Mapped[str | None]
    affected_milestone: Mapped[str | None]
    resolution_plan: Mapped[str | None]
    due_date: Mapped[date | None]
    actual_resolution_date: Mapped[date | None]
    status: Mapped[str]
    escalation_level: Mapped[str | None]
    escalation_date: Mapped[date | None]
    resolution_summary: Mapped[str | None]
    lessons_learned: Mapped[str | None]
    closure_date: Mapped[date | None]
    remarks: Mapped[str | None]
    last_review_date: Mapped[date | None]
    next_review_date: Mapped[date | None]


class DependencyLog(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "dependency_log"

    dependency_code: Mapped[str] = mapped_column(unique=True)
    project_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("projects.id", ondelete="CASCADE"))

    dependency_title: Mapped[str]
    description: Mapped[str | None]
    dependency_type: Mapped[str | None]
    category: Mapped[str | None]
    depends_on: Mapped[str | None]
    related_task_milestone: Mapped[str | None]
    required_by_date: Mapped[date | None]
    owner: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    dependency_status: Mapped[str]
    criticality: Mapped[str | None]
    impact_if_delayed: Mapped[str | None]
    probability_of_delay: Mapped[str | None]
    mitigation_plan: Mapped[str | None]
    escalation_required: Mapped[bool]
    escalation_level: Mapped[str | None]
    actual_completion_date: Mapped[date | None]
    last_updated: Mapped[date | None]
    remarks: Mapped[str | None]
    last_review_date: Mapped[date | None]
    next_review_date: Mapped[date | None]


class AssumptionLog(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "assumption_log"

    assumption_code: Mapped[str] = mapped_column(unique=True)
    project_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("projects.id", ondelete="CASCADE"))

    title: Mapped[str]
    detailed_description: Mapped[str | None]
    category: Mapped[str | None]
    raised_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    raised_date: Mapped[date | None]
    owner: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    dependency_reference: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("dependency_log.id"))
    impact_if_invalid: Mapped[str | None]
    probability_of_failure: Mapped[str | None]
    impact_rating: Mapped[str | None]
    validation_date: Mapped[date | None]
    validation_status: Mapped[str]
    mitigation_plan: Mapped[str | None]
    contingency_plan: Mapped[str | None]
    current_status: Mapped[str]
    last_updated: Mapped[date | None]
    remarks: Mapped[str | None]


class OpportunityLog(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "opportunity_log"

    opportunity_code: Mapped[str] = mapped_column(unique=True)
    project_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("projects.id", ondelete="CASCADE"))

    opportunity_title: Mapped[str]
    opportunity_description: Mapped[str | None]
    category: Mapped[str | None]
    identified_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    identified_date: Mapped[date | None]
    opportunity_owner: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    impact: Mapped[str | None]
    expected_benefit: Mapped[str | None]
    estimated_benefit: Mapped[Decimal | None] = mapped_column(Numeric)
    benefit_type: Mapped[str | None]
    exploitation_strategy: Mapped[str | None]
    action_plan: Mapped[str | None]
    target_implementation_date: Mapped[date | None]
    status: Mapped[str]
    approval_required: Mapped[bool]
    approved_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    actual_benefit: Mapped[Decimal | None] = mapped_column(Numeric)
    closure_date: Mapped[date | None]
    remarks: Mapped[str | None]
    last_review_date: Mapped[date | None]
    next_review_date: Mapped[date | None]

File: frontend/src/lib/api/raid.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api, type Page } from "./client";

// All five RAID logs share one CRUD shape under /projects/{project_id}/{prefix}
// (see backend/app/api/v1/endpoints/raid.py's build_raid_router), including
// PUT/DELETE per item — these four generic hooks back every entity below.
function useRaidList<T>(projectId: string | null, prefix: string) {
  return useQuery({
    queryKey: ["raid", prefix, projectId],
    queryFn: () => api.get<Page<T>>(`/projects/${projectId}/${prefix}?limit=200`),
    select: (page) => page.items,
    enabled: !!projectId,
  });
}

function useRaidCreate<TPayload, TRead>(projectId: string | null, prefix: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: TPayload) => api.post<TRead>(`/projects/${projectId}/${prefix}`, payload),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["raid", prefix, projectId] });
    },
  });
}

function useRaidUpdate<TPayload, TRead>(projectId: string | null, prefix: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, payload }: { id: string; payload: TPayload }) =>
      api.put<TRead>(`/projects/${projectId}/${prefix}/${id}`, payload),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["raid", prefix, projectId] });
    },
  });
}

function useRaidDelete(projectId: string | null, prefix: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => api.delete(`/projects/${projectId}/${prefix}/${id}`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["raid", prefix, projectId] });
    },
  });
}

// --- Risk ---

export type RiskLog = {
  id: string;
  risk_code: string;
  project_id: string;
  risk_title: string;
  risk_description: string | null;
  risk_category: string | null;
  risk_type: string | null;
  identified_by: string | null;
  identified_date: string | null;
  risk_owner: string | null;
  trigger_event: string | null;
  probability: string | null;
  impact: string | null;
  risk_score: number | null;
  severity: string | null;
  affected_deliverables: string | null;
  affected_milestone: string | null;
  response_strategy: string | null;
  mitigation_plan: string | null;
  contingency_plan: string | null;
  residual_risk: string | null;
  target_resolution_date: string | null;
  current_status: string;
  escalation_required: boolean;
  escalated_to: string | null;
  last_review_date: string | null;
  next_review_date: string | null;
  remarks: string | null;
};

export type RiskLogPayload = Partial<Omit<RiskLog, "id" | "risk_code" | "project_id" | "current_status">>;

export const useRisks = (projectId: string | null) => useRaidList<RiskLog>(projectId, "risks");
export const useCreateRisk = (projectId: string | null) =>
  useRaidCreate<RiskLogPayload, RiskLog>(projectId, "risks");
export const useUpdateRisk = (projectId: string | null) =>
  useRaidUpdate<RiskLogPayload, RiskLog>(projectId, "risks");
export const useDeleteRisk = (projectId: string | null) => useRaidDelete(projectId, "risks");

// --- Issue ---

export type IssueLog = {
  id: string;
  issue_code: string;
  project_id: string;
  issue_title: string;
  issue_description: string | null;
  issue_category: string | null;
  priority: string | null;
  severity: string | null;
  raised_by: string | null;
  raised_date: string | null;
  assigned_to: string | null;
  root_cause: string | null;
  business_impact: string | null;
  affected_deliverables: string | null;
  affected_milestone: string | null;
  resolution_plan: string | null;
  due_date: string | null;
  actual_resolution_date: string | null;
  status: string;
  escalation_level: string | null;
  escalation_date: string | null;
  resolution_summary: string | null;
  lessons_learned: string | null;
  closure_date: string | null;
  remarks: string | null;
  last_review_date: string | null;
  next_review_date: string | null;
};

export type IssueLogPayload = Partial<Omit<IssueLog, "id" | "issue_code" | "project_id" | "status">>;

export const useIssues = (projectId: string | null) => useRaidList<IssueLog>(projectId, "issues");
export const useCreateIssue = (projectId: string | null) =>
  useRaidCreate<IssueLogPayload, IssueLog>(projectId, "issues");
export const useUpdateIssue = (projectId: string | null) =>
  useRaidUpdate<IssueLogPayload, IssueLog>(projectId, "issues");
export const useDeleteIssue = (projectId: string | null) => useRaidDelete(projectId, "issues");

// --- Dependency ---

export type DependencyLog = {
  id: string;
  dependency_code: string;
  project_id: string;
  dependency_title: string;
  description: string | null;
  dependency_type: string | null;
  category: string | null;
  depends_on: string | null;
  related_task_milestone: string | null;
  required_by_date: string | null;
  owner: string | null;
  dependency_status: string;
  criticality: string | null;
  impact_if_delayed: string | null;
  probability_of_delay: string | null;
  mitigation_plan: string | null;
  escalation_required: boolean;
  escalation_level: string | null;
  actual_completion_date: string | null;
  last_updated: string | null;
  remarks: string | null;
  last_review_date: string | null;
  next_review_date: string | null;
};

export type DependencyLogPayload = Partial<
  Omit<DependencyLog, "id" | "dependency_code" | "project_id" | "dependency_status">
>;

export const useDependencies = (projectId: string | null) => useRaidList<DependencyLog>(projectId, "dependencies");
export const useCreateDependency = (projectId: string | null) =>
  useRaidCreate<DependencyLogPayload, DependencyLog>(projectId, "dependencies");
export const useUpdateDependency = (projectId: string | null) =>
  useRaidUpdate<DependencyLogPayload, DependencyLog>(projectId, "dependencies");
export const useDeleteDependency = (projectId: string | null) => useRaidDelete(projectId, "dependencies");

// --- Assumption ---

export type AssumptionLog = {
  id: string;
  assumption_code: string;
  project_id: string;
  title: string;
  detailed_description: string | null;
  category: string | null;
  raised_by: string | null;
  raised_date: string | null;
  owner: string | null;
  dependency_reference: string | null;
  impact_if_invalid: string | null;
  probability_of_failure: string | null;
  impact_rating: string | null;
  validation_date: string | null;
  validation_status: string;
  mitigation_plan: string | null;
  contingency_plan: string | null;
  current_status: string;
  last_updated: string | null;
  remarks: string | null;
};

export type AssumptionLogPayload = Partial<
  Omit<AssumptionLog, "id" | "assumption_code" | "project_id" | "validation_status" | "current_status">
>;

export const useAssumptions = (projectId: string | null) => useRaidList<AssumptionLog>(projectId, "assumptions");
export const useCreateAssumption = (projectId: string | null) =>
  useRaidCreate<AssumptionLogPayload, AssumptionLog>(projectId, "assumptions");
export const useUpdateAssumption = (projectId: string | null) =>
  useRaidUpdate<AssumptionLogPayload, AssumptionLog>(projectId, "assumptions");
export const useDeleteAssumption = (projectId: string | null) => useRaidDelete(projectId, "assumptions");

// --- Opportunity ---

export type OpportunityLog = {
  id: string;
  opportunity_code: string;
  project_id: string;
  opportunity_title: string;
  opportunity_description: string | null;
  category: string | null;
  identified_by: string | null;
  identified_date: string | null;
  opportunity_owner: string | null;
  impact: string | null;
  expected_benefit: string | null;
  estimated_benefit: string | null;
  benefit_type: string | null;
  exploitation_strategy: string | null;
  action_plan: string | null;
  target_implementation_date: string | null;
  status: string;
  approval_required: boolean;
  approved_by: string | null;
  actual_benefit: string | null;
  closure_date: string | null;
  remarks: string | null;
  last_review_date: string | null;
  next_review_date: string | null;
};

export type OpportunityLogPayload = Partial<
  Omit<OpportunityLog, "id" | "opportunity_code" | "project_id" | "status">
>;

export const useOpportunities = (projectId: string | null) => useRaidList<OpportunityLog>(projectId, "opportunities");
export const useCreateOpportunity = (projectId: string | null) =>
  useRaidCreate<OpportunityLogPayload, OpportunityLog>(projectId, "opportunities");
export const useUpdateOpportunity = (projectId: string | null) =>
  useRaidUpdate<OpportunityLogPayload, OpportunityLog>(projectId, "opportunities");
export const useDeleteOpportunity = (projectId: string | null) => useRaidDelete(projectId, "opportunities");
